File: py/branchio_sdk/features.py
# Branchio SDK feature factory

from branchio_sdk.feature.base_feature import BranchioBaseFeature
from branchio_sdk.feature.debug_feature import BranchioDebugFeature
from branchio_sdk.feature.idempotency_feature import BranchioIdempotencyFeature
from branchio_sdk.feature.metrics_feature import BranchioMetricsFeature
from branchio_sdk.feature.paging_feature import BranchioPagingFeature
from branchio_sdk.feature.ratelimit_feature import BranchioRatelimitFeature
from branchio_sdk.feature.retry_feature import BranchioRetryFeature
from branchio_sdk.feature.test_feature import BranchioTestFeature
from branchio_sdk.feature.timeout_feature import BranchioTimeoutFeature


_FEATURES = {
    "base": lambda: BranchioBaseFeature(),
    "debug": lambda: BranchioDebugFeature(),
    "idempotency": lambda: BranchioIdempotencyFeature(),
    "metrics": lambda: BranchioMetricsFeature(),
    "paging": lambda: BranchioPagingFeature(),
    "ratelimit": lambda: BranchioRatelimitFeature(),
    "retry": lambda: BranchioRetryFeature(),
    "test": lambda: BranchioTestFeature(),
    "timeout": lambda: BranchioTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// Branchio SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class BranchioFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new BranchioBaseFeature();
            case "debug":
                return new BranchioDebugFeature();
            case "idempotency":
                return new BranchioIdempotencyFeature();
            case "metrics":
                return new BranchioMetricsFeature();
            case "paging":
                return new BranchioPagingFeature();
            case "ratelimit":
                return new BranchioRatelimitFeature();
            case "retry":
                return new BranchioRetryFeature();
            case "test":
                return new BranchioTestFeature();
            case "timeout":
                return new BranchioTimeoutFeature();
            default:
                return new BranchioBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
